src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1877,175 +2095,363 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 		 */
 		if (hasAggregates && !hasWindowFunction)
 		{
-			WorkerAggregateWalker((Node *) originalExpression, walkerContext);
+			WorkerAggregateWalker((Node *) originalExpression, workerAggContext);
 
-			newExpressionList = walkerContext->expressionList;
-			queryHasAggregates = true;
+			newExpressionList = workerAggContext->expressionList;
 		}
 		else
 		{
 			newExpressionList = list_make1(originalExpression);
 		}
 
-		createNewGroupByClause = walkerContext->createGroupByClause;
-
 		ExpandWorkerTargetEntry(newExpressionList, originalTargetEntry,
-								createNewGroupByClause, &newTargetEntryList,
-								&targetProjectionNumber, &groupClauseList,
-								&nextSortGroupRefIndex);
+								workerAggContext->createGroupByClause,
+								queryTargetList->targetEntryList,
+								queryTargetList->targetProjectionNumber,
+								queryGroupClause->groupClauseList,
+								queryGroupClause->nextSortGroupRefIndex);
```

Comment:
Could we pass on the structs to ExpandWorkerTargetEntry as well?

Reply:
Yes, makes sense. Even to one lower level function `AppendTargetEntryToGroupClause`